### PR TITLE
Fix a problem with connection waiters that are never awaited

### PR DESCRIPTION
**Pull request by illia-v**
## What do these changes do?
When a connector waits for multiple available connections with the same key, the described race condition can happen.

Let's say that we want to create 3 connections but there a no available ones due to a limit. Then:
1. The first future [is added](https://github.com/aio-libs/aiohttp/blob/ea8f35e80539931e864922c2214bc4a3ec7d9cf4/aiohttp/connector.py#L465), `BaseConnector.connect` [sets](https://github.com/aio-libs/aiohttp/blob/ea8f35e80539931e864922c2214bc4a3ec7d9cf4/aiohttp/connector.py#L464) a deque 1 containing the future to a local `waiters` variable, the future is awaited, the context is switched. 
2. The second future is added, `BaseConnector.connect` sets the deque 1 containing first and second futures to `waiters`, the second future is awaited, the context is switched.
3. One of the previous connections is closed, the first future [is popped](https://github.com/aio-libs/aiohttp/blob/ea8f35e80539931e864922c2214bc4a3ec7d9cf4/aiohttp/connector.py#L578) from the deque 1, its result [is set](https://github.com/aio-libs/aiohttp/blob/ea8f35e80539931e864922c2214bc4a3ec7d9cf4/aiohttp/connector.py#L580), the context is switched.
4. Another previous connection is closed, the second future is popped from the deque 1 (the deque is empty), its result is set, the context is switched.
5. We returned to the context from the first step, the deque 1 [is deleted](https://github.com/aio-libs/aiohttp/blob/ea8f35e80539931e864922c2214bc4a3ec7d9cf4/aiohttp/connector.py#L485) from `BaseConnector._waiters`, ..., the context is switched.
6. The third future is added, `BaseConnector.connect` sets a new deque 2 containing the third future to `waiters`, the third future is awaited, the context is switched.
7. Asyncio returned to the context from the second step, it checks that the local `waiters` variable containing the deque 1 is empty, and deletes `self._waiters[key]`. But `self._waiters[key]` is the deque 2 at the moment. Therefore, the waiter represented as the third future will be never released. And the program will hang if no `connect` timeout is set.

## Are there changes in behavior for the user?
No

## Related issue number
Maybe #4258 and https://github.com/aio-libs/aiobotocore/issues/738

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by illia-v** [comment]
```python
In [1]: from collections import defaultdict, deque                                                                                                                                                                                                                                                        

In [2]: dictionary = defaultdict(deque)                                                                                                                                                                                                                                                                   

In [3]: dq = dictionary['key']                                                                                                                                                                                                                                                                            

In [4]: dq.append('future')                                                                                                                                                                                                                                                                               

In [5]: dq is dictionary['key']                                                                                                                                                                                                                                                                           
Out[5]: True

In [6]: del dictionary['key']                                                                                                                                                                                                                                                                             

In [7]: dq                                                                                                                                                                                                                                                                                                
Out[7]: deque(['future'])

In [8]: dictionary                                                                                                                                                                                                                                                                                        
Out[8]: defaultdict(collections.deque, {})

In [9]: dictionary['key'].append('future')                                                                                                                                                                                                                                                                

In [10]: dictionary                                                                                                                                                                                                                                                                                       
Out[10]: defaultdict(collections.deque, {'key': deque(['future'])})

In [11]: dq is dictionary['key']                                                                                                                                                                                                                                                                          
Out[11]: False
```

**Comment by thehesiod** [comment]
this logic is getting impossible to reason with, how about something like this: https://github.com/thehesiod/aiohttp/pull/4  The issue I have right now is that `_release_waiter` is not async and doesn't appear to be easy to make async

**Comment by thehesiod** [comment]
actually I think I came up with a fix in my PR

**Comment by illia-v** [comment]
@thehesiod the race condition described in the first comment caused the https://github.com/aio-libs/aiobotocore/issues/738#issuecomment-581441776 problem. When I was debugging it, it was pretty clear that a different deque with a same key was deleted from the dictionary. I checked twice, and therefore I don't agree that it is impossible :)

I'll take a look at your PR when I can

**Comment by illia-v** [comment]
@thehesiod I looked at your pull request and added two minor comments.

Af for me your solution looks better than the current implementation. Also, it seems to fix the problem mentioned in this PR too, I haven't checked it though.

If your changes are merged before the next release, I'm OK with closing this PR.
Otherwise, I'd like this one to be merged before the next release to fix the problem because it changes only one line and should not create any regressions

**Comment by thehesiod** [comment]
@illia-v I didn't open my PR against aiohttp, up to maintainers of aiohttp to see if they want that or not ( @asvetlov )

File: aiohttp/connector.py [hunk 1]
```diff
@@ -480,7 +480,10 @@ async def connect(self, req: 'ClientRequest',
 
                 raise e
             finally:
-                if not waiters:
+                # `waiters` may be deleted from `self._waiters` by another
+                # coroutine, and `self._waiters[key]` may contain another
+                # deque, the later should not be deleted.
```
Comment (pfreixes) [inline aiohttp/connector.py]: I guess that the issue is that we are caching the `self._waiters[key]` at the local level by using the local variable `waiters` what about simply doing a check with using the none cached version, like:

```python
if key in self._waiters and not self._waiters[key]:
    ...
```
Maybe it's more clear.

WDYT?

Comment (illia-v) [reply aiohttp/connector.py]: It's a good idea. Thanks!

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
LGTM, thanks!

**Comment by thehesiod** [comment]
@asvetlov welcome back :)

**Comment by asvetlov** [comment]
Thanks, @thehesiod

**Pull request merged**